control/utils/graph.py

```
@@ -0,0 +1,327 @@
+    """
+    Reads a single log file, processes sensor data, and calculates timestamps, RPMs, and source line numbers.
+    """
+    timestamps_s = []
+    rpms = []
+    source_line_numbers = []
+    current_total_time_s = 0.0
+    logging_active = False
+    processed_data_for_this_file = False
+
+    try:
+        with open(filepath, 'r') as f:
```

Pylint (3.3.7)

[warning] Using open without explicitly specifying an encoding

(W1514)